crates/ironrdp-web/src/session.rs

```diff
@@ -492,18 +477,120 @@ impl Session {
+                                &JsValue::from_str("default"),
+                                &JsValue::UNDEFINED,
+                                &JsValue::UNDEFINED,
+                            )
+                            .map_err(|e| anyhow::Error::msg(format!("set cursor style callback failed: {e:?}")))?;
                     }
                     ActiveStageOutput::PointerHidden => {
                         let _ret = self
-                            .hide_pointer_callback
-                            .call0(&self.hide_pointer_callback_context)
-                            .map_err(|e| anyhow::Error::msg(format!("hide pointer callback failed: {e:?}")))?;
+                            .set_cursor_style_callback
+                            .call3(
+                                &self.set_cursor_style_callback_context,
+                                &JsValue::from_str("none"),
+                                &JsValue::UNDEFINED,
+                                &JsValue::UNDEFINED,
+                            )
+                            .map_err(|e| anyhow::Error::msg(format!("set cursor style callback failed: {e:?}")))?;
                     }
                     ActiveStageOutput::PointerPosition { .. } => {
-                        // Not applicable for web
+                        // Not applicable for web.
+                    }
+                    ActiveStageOutput::NativePointerUpdate(pointer) => {
+                        // Maximum allowed cursor size for browsers is 32x32, because bigger sizes
+                        // will cause the following issues:
+                        // - 128x128 and bigger cursors are not supported in browsers.
+                        // - 32x32 and bigger cursors will default to the system cursor if their
+                        //   sprite does not fit in the browser's viewport, introducing anabrupt
+                        //   cursor style changes when the cursor is moved to the edge of the
+                        //   browser window.
+                        //
+                        // Therefore, we need to scale the cursor sprite down to 32x32 if it is
+                        // bigger than that.
+                        const MAX_CURSOR_SIZE: u16 = 32;
+                        // INVARIANT: 0 < scale <= 1.0
+                        let scale = if pointer.width >= pointer.height && pointer.width > MAX_CURSOR_SIZE {
+                            Some(f64::from(MAX_CURSOR_SIZE) / f64::from(pointer.width))
+                        } else if pointer.height > MAX_CURSOR_SIZE {
+                            Some(f64::from(MAX_CURSOR_SIZE) / f64::from(pointer.height))
+                        } else {
+                            None
+                        };
+
+                        let (png_width, png_height, hot_spot_x, hot_spot_y, rgba_buffer) = if let Some(scale) = scale {
+                            // Per invariants: Following conversions are losslesss.
```

Comment:
Question: are they really lossless? For instance, when the cursor is not a square, either `width * scale` or `height * scale` may result in something that is not an integer. Did you mean that the cast will never "saturate"?

Reply:
Ah, yes, the working is a bit off. Sure, that is about saturation

Reply:
Sounds good! In that case, the naming for `cast_f64_to_u16_lossy` should be changed too